### ISSUE TITLE
linter: Add help or notes to diagnostics only containing a message

### ISSUE THREAD

**Issue by camchenry** [oxc-project/oxc#19121]
> [!NOTE]
> The majority of this issue was generated automatically, so it may contain false positives. Be sure to look at the rule output carefully to ensure it actually doesn't contain any help text or notes.

It is helpful when diagnostics include help text or notes that provide context on how to fix an issue and why it occurred. In general, almost all linter diagnostics should include text like this in order to be as helpful as possible.

If you aren't sure if a diagnostic message is acceptable, feel free to reach out in Discord in `#linter` to discuss.

## Diagnostics missing `with_help` and `with_note`

> Generated by `ast-grep scan --rule diagnostic-missing-help-or-note.yml`

<details>

```yml
id: diagnostic-missing-help-or-note
language: Rust
severity: warning
message: OxcDiagnostic is missing a call to `with_help` or `with_note`. All linter diagnostics should provide actionable help or explanatory notes.
note: Add `.with_help("...")` or `.with_note("...")` to the diagnostic chain.
files:
  - "crates/oxc_linter/src/rules/**/*.rs"
rule:
  any:
    - pattern: OxcDiagnostic::warn($$$)
    - pattern: OxcDiagnostic::error($$$)
  not:
    any:
      # Directly chained: OxcDiagnostic::warn(...).with_help/with_note(...)
      - inside:
          any:
            - pattern: $A.with_help($$$)
            - pattern: $A.with_note($$$)
          stopBy: end
      # Variable assignment pattern: the OxcDiagnostic is inside a function
      # that contains with_help or with_note calls somewhere
      - inside:
          all:
            - kind: function_item
            - has:
                kind: field_identifier
                regex: "^with_(help|note)$"
                stopBy: end
          stopBy: end
```

</details>

### `eslint`

<details>

- [x] `crates/oxc_linter/src/rules/eslint/new_cap.rs:19` — msg
- [ ] `crates/oxc_linter/src/rules/eslint/no_async_promise_executor.rs:12` — Promise executor functions should not be `async`.
- [ ] `crates/oxc_linter/src/rules/eslint/no_await_in_loop.rs:12` — Unexpected `await` inside a loop.
- [ ] `crates/oxc_linter/src/rules/eslint/no_case_declarations.rs:12` — Unexpected lexical declaration in case block.
- [ ] `crates/oxc_linter/src/rules/eslint/no_class_assign.rs:10` — Unexpected re-assignment of class {name}
- [ ] `crates/oxc_linter/src/rules/eslint/no_const_assign.rs:10` — Unexpected re-assignment of `const` variable {name}.
- [ ] `crates/oxc_linter/src/rules/eslint/no_constructor_return.rs:13` — Unexpected return statement in constructor.
- [ ] `crates/oxc_linter/src/rules/eslint/no_control_regex.rs:37` — if count > 1 {
        "Unexpected control characters"
    } else {
        "Unexpected control character"
    }
- [ ] `crates/oxc_linter/src/rules/eslint/no_delete_var.rs:10` — Variables should not be deleted
- [ ] `crates/oxc_linter/src/rules/eslint/no_eval.rs:17` — eval can be harmful.
- [ ] `crates/oxc_linter/src/rules/eslint/no_extend_native.rs:115` — format!(
                            "{name} prototype is read-only, properties should not be added."
                        
- [ ] `crates/oxc_linter/src/rules/eslint/no_extend_native.rs:126` — format!(
                            "{name} prototype is read-only, properties should not be added."
                        
- [ ] `crates/oxc_linter/src/rules/eslint/no_fallthrough.rs:26` — Expected a `break` statement before `case`.
- [ ] `crates/oxc_linter/src/rules/eslint/no_fallthrough.rs:30` — Expected a `break` statement before `default`.
- [ ] `crates/oxc_linter/src/rules/eslint/no_fallthrough.rs:34` — 
        "Found a comment that would permit fallthrough, but case cannot fall through.",
    
- [ ] `crates/oxc_linter/src/rules/eslint/no_func_assign.rs:10` — '{name}' is a function.
- [ ] `crates/oxc_linter/src/rules/eslint/no_global_assign.rs:14` — Read-only global '{global_name}' should not be modified.
- [ ] `crates/oxc_linter/src/rules/eslint/no_label_var.rs:9` — Found identifier '{name}' with the same name as a label.
- [ ] `crates/oxc_linter/src/rules/eslint/no_labels.rs:19` — message
- [ ] `crates/oxc_linter/src/rules/eslint/no_lone_blocks.rs:10` — Block is unnecessary.
- [ ] `crates/oxc_linter/src/rules/eslint/no_lone_blocks.rs:14` — Nested block is redundant.
- [ ] `crates/oxc_linter/src/rules/eslint/no_loss_of_precision.rs:12` — This number literal will lose precision at runtime.
- [ ] `crates/oxc_linter/src/rules/eslint/no_magic_numbers.rs:21` — Number constants declarations must use 'const'.
- [ ] `crates/oxc_linter/src/rules/eslint/no_magic_numbers.rs:25` — No magic number: {raw}
- [ ] `crates/oxc_linter/src/rules/eslint/no_magic_numbers.rs:80` — 
                    "Expecting object for eslint/no-magic-numbers configuration",
                
- [ ] `crates/oxc_linter/src/rules/eslint/no_misleading_character_class.rs:20` — Unexpected surrogate pair in character class.
- [ ] `crates/oxc_linter/src/rules/eslint/no_misleading_character_class.rs:24` — Unexpected combining class in character class.
- [ ] `crates/oxc_linter/src/rules/eslint/no_misleading_character_class.rs:28` — Unexpected emoji modifier in character class.
- [ ] `crates/oxc_linter/src/rules/eslint/no_misleading_character_class.rs:32` — Unexpected regional indicator in character class.
- [ ] `crates/oxc_linter/src/rules/eslint/no_misleading_character_class.rs:36` — Unexpected joined character sequence in character class.
- [ ] `crates/oxc_linter/src/rules/eslint/no_multi_str.rs:9` — Unexpected multi string.
- [ ] `crates/oxc_linter/src/rules/eslint/no_nested_ternary.rs:10` — Do not nest ternary expressions.
- [ ] `crates/oxc_linter/src/rules/eslint/no_new.rs:10` — Do not use 'new' for side effects.
- [ ] `crates/oxc_linter/src/rules/eslint/no_new_func.rs:10` — The Function constructor is eval.
- [ ] `crates/oxc_linter/src/rules/eslint/no_new_native_nonconstructor.rs:9` — `{fn_name}` cannot be called as a constructor.
- [ ] `crates/oxc_linter/src/rules/eslint/no_nonoctal_decimal_escape.rs:15` — Don't use '{escape_sequence}' escape sequence.
- [ ] `crates/oxc_linter/src/rules/eslint/no_param_reassign.rs:25` — Assignment to function parameter '{name}'.
- [ ] `crates/oxc_linter/src/rules/eslint/no_param_reassign.rs:29` — Assignment to property of function parameter '{name}'.
- [ ] `crates/oxc_linter/src/rules/eslint/no_redeclare.rs:15` — '{name}' is already defined.
- [ ] `crates/oxc_linter/src/rules/eslint/no_redeclare.rs:22` — '{name}' is already defined as a built-in global variable.
- [ ] `crates/oxc_linter/src/rules/eslint/no_restricted_globals.rs:18` — warn_text
- [ ] `crates/oxc_linter/src/rules/eslint/no_self_assign.rs:23` — this expression is assigned to itself
- [ ] `crates/oxc_linter/src/rules/eslint/no_setter_return.rs:9` — Setter cannot return a value
- [ ] `crates/oxc_linter/src/rules/eslint/no_throw_literal.rs:12` — message
- [ ] `crates/oxc_linter/src/rules/eslint/no_undef.rs:16` — '{name}' is not defined.
- [ ] `crates/oxc_linter/src/rules/eslint/no_undefined.rs:12` — Unexpected use of `undefined`
- [ ] `crates/oxc_linter/src/rules/eslint/no_unreachable.rs:17` — Unreachable code.
- [ ] `crates/oxc_linter/src/rules/eslint/no_unused_private_class_members.rs:12` — '{name}' is defined but never used.
- [ ] `crates/oxc_linter/src/rules/eslint/no_unused_vars/options.rs:455` — Invalid '{option_name}' option for no-unused-vars: {message}
- [ ] `crates/oxc_linter/src/rules/eslint/no_unused_vars/options.rs:530` — Invalid '{name}' option for no-unused-vars: {err}
- [ ] `crates/oxc_linter/src/rules/eslint/no_unused_vars/options.rs:618` — format!(
                "Invalid 'vars' option for no-unused-vars: Expected a string or an object, got {config}"
- [ ] `crates/oxc_linter/src/rules/eslint/no_useless_backreference.rs:19` — Backreference '{back_reference}' will be ignored. It references group '{group}' from within that group.
- [ ] `crates/oxc_linter/src/rules/eslint/no_useless_backreference.rs:20` — Backreference '{back_reference}' will be ignored. It references group '{group}' which is in another alternative.
- [ ] `crates/oxc_linter/src/rules/eslint/no_useless_backreference.rs:21` — Backreference '{back_reference}' will be ignored. It references group '{group}' which appears later in the pattern.
- [ ] `crates/oxc_linter/src/rules/eslint/no_useless_backreference.rs:22` — Backreference '{back_reference}' will be ignored. It references group '{group}' which appears before in the same lookbehind.
- [ ] `crates/oxc_linter/src/rules/eslint/no_useless_backreference.rs:23` — Backreference '{back_reference}' will be ignored. It references group '{group}' which is in a negative lookaround.
- [ ] `crates/oxc_linter/src/rules/eslint/no_useless_catch.rs:12` — Unnecessary try/catch wrapper
- [ ] `crates/oxc_linter/src/rules/eslint/no_useless_catch.rs:17` — Unnecessary catch clause
- [ ] `crates/oxc_linter/src/rules/eslint/operator_assignment.rs:24` — msg
- [ ] `crates/oxc_linter/src/rules/eslint/prefer_exponentiation_operator.rs:16` — Prefer `**` over `Math.pow`.
- [ ] `crates/oxc_linter/src/rules/eslint/prefer_numeric_literals.rs:15` — Use {prefix_name} literals instead of parseInt().
- [ ] `crates/oxc_linter/src/rules/eslint/prefer_object_has_own.rs:12` — 
        "Disallow use of `Object.prototype.hasOwnProperty.call(
- [ ] `crates/oxc_linter/src/rules/eslint/prefer_promise_reject_errors.rs:21` — Expected the Promise rejection reason to be an Error
- [ ] `crates/oxc_linter/src/rules/eslint/prefer_rest_params.rs:8` — Use the rest parameters instead of `arguments`.
- [ ] `crates/oxc_linter/src/rules/eslint/prefer_spread.rs:12` — Use spread operators instead of `.apply()`.
- [ ] `crates/oxc_linter/src/rules/eslint/radix.rs:19` — Missing parameters.
- [ ] `crates/oxc_linter/src/rules/eslint/radix.rs:29` — Redundant radix parameter.
- [ ] `crates/oxc_linter/src/rules/eslint/radix.rs:33` — Invalid radix parameter, must be an integer between 2 and 36.
- [ ] `crates/oxc_linter/src/rules/eslint/require_yield.rs:9` — This generator function does not have `yield`
- [ ] `crates/oxc_linter/src/rules/eslint/sort_imports.rs:26` — Expected '{curr_kind}' syntax before '{prev_kind}' syntax.
- [ ] `crates/oxc_linter/src/rules/eslint/sort_imports.rs:31` — Imports should be sorted alphabetically.
- [ ] `crates/oxc_linter/src/rules/eslint/sort_imports.rs:35` — format!(
        "Member '{name}' of the import declaration should be sorted alphabetically."
- [ ] `crates/oxc_linter/src/rules/eslint/sort_keys.rs:61` — Object keys should be sorted
- [ ] `crates/oxc_linter/src/rules/eslint/sort_vars.rs:16` — Variable declarations should be sorted
- [ ] `crates/oxc_linter/src/rules/eslint/symbol_description.rs:12` — Expected Symbol to have a description.

</details>

### `import`

<details>

- [ ] `crates/oxc_linter/src/rules/import/export.rs:12` — No named exports found in module '{module_name}'
- [ ] `crates/oxc_linter/src/rules/import/export.rs:95` — Multiple exports of name '{name}'.
- [ ] `crates/oxc_linter/src/rules/import/exports_last.rs:11` — Export statements should appear at the end of the file
- [ ] `crates/oxc_linter/src/rules/import/namespace.rs:21` — format!(
        "{specifier_name:?} not found in imported namespace {namespace_name:?}."
- [ ] `crates/oxc_linter/src/rules/import/namespace.rs:32` — format!(
        "{specifier_name:?} not found in deeply imported namespace {namespace_name:?}."
- [ ] `crates/oxc_linter/src/rules/import/namespace.rs:39` — format!(
        "Unable to validate computed reference to imported namespace {namespace_name:?}."
- [ ] `crates/oxc_linter/src/rules/import/namespace.rs:46` — Assignment to member of namespace {namespace_name:?}.'
- [ ] `crates/oxc_linter/src/rules/import/no_absolute_path.rs:22` — Do not import modules using an absolute path
- [ ] `crates/oxc_linter/src/rules/import/no_anonymous_default_export.rs:20` — msg
- [ ] `crates/oxc_linter/src/rules/import/no_default_export.rs:8` — Prefer named exports
- [ ] `crates/oxc_linter/src/rules/import/no_mutable_exports.rs:16` — Exporting mutable '{kind_str}' binding, use 'const' instead.
- [ ] `crates/oxc_linter/src/rules/import/no_nodejs_modules.rs:20` — Do not import Node.js builtin module `{module_name}`
- [ ] `crates/oxc_linter/src/rules/import/no_self_import.rs:8` — A module importing itself is not allowed

</details>

### `jest`

<details>

- [ ] `crates/oxc_linter/src/rules/jest/no_conditional_in_test.rs:13` — Avoid having conditionals in tests.
- [ ] `crates/oxc_linter/src/rules/jest/no_confusing_set_timeout.rs:15` — `jest.setTimeout` should only be called in a global scope
- [ ] `crates/oxc_linter/src/rules/jest/no_confusing_set_timeout.rs:26` — `jest.setTimeout` should be placed before any other jest methods.
- [ ] `crates/oxc_linter/src/rules/jest/no_hooks.rs:16` — Do not use setup or teardown hooks.
- [ ] `crates/oxc_linter/src/rules/jest/no_restricted_jest_methods.rs:16` — Use of `{method_name}` is not allowed
- [ ] `crates/oxc_linter/src/rules/jest/no_restricted_jest_methods.rs:20` — message.to_string(
- [ ] `crates/oxc_linter/src/rules/jest/no_restricted_matchers.rs:20` — Use of `{chain_call}` is disallowed
- [ ] `crates/oxc_linter/src/rules/jest/no_test_return_statement.rs:18` — Jest tests should not return a value

</details>

### `oxc`

- [ ] `crates/oxc_linter/src/rules/oxc/no_rest_spread_properties.rs:18` — {spread_kind} are not allowed. {message_suffix}

### `promise`

<details>

- [ ] `crates/oxc_linter/src/rules/promise/always_return.rs:29` — Each then() should return a value or throw
- [ ] `crates/oxc_linter/src/rules/promise/avoid_new.rs:9` — Avoid creating new promises
- [ ] `crates/oxc_linter/src/rules/promise/no_multiple_resolved.rs:26` — format!(
        "Promise should not be resolved multiple times. Promise is already resolved on line {line}."
- [ ] `crates/oxc_linter/src/rules/promise/no_multiple_resolved.rs:33` — Promise should not be resolved multiple times. Promise is potentially resolved on line {line}.
- [ ] `crates/oxc_linter/src/rules/promise/param_names.rs:14` — format!(
        "Promise constructor parameters must be named to match `{pattern}`"
- [ ] `crates/oxc_linter/src/rules/promise/prefer_await_to_callbacks.rs:13` — Prefer `async`/`await` to the callback pattern
- [ ] `crates/oxc_linter/src/rules/promise/prefer_await_to_then.rs:9` — Prefer await to then()/catch()/finally()
- [ ] `crates/oxc_linter/src/rules/promise/spec_only.rs:16` — Avoid using non-standard `Promise.{prop_name}` method.
- [ ] `crates/oxc_linter/src/rules/promise/valid_params.rs:13` — format!(
        "`Promise.{prop_name}(
- [ ] `crates/oxc_linter/src/rules/promise/valid_params.rs:24` — format!(
        "`Promise.{prop_name}(
- [ ] `crates/oxc_linter/src/rules/promise/valid_params.rs:31` — format!(
        "`Promise.{prop_name}(

</details>

### `react`

<details>

- [ ] `crates/oxc_linter/src/rules/react/forbid_dom_props.rs:21` — message.clone(
- [ ] `crates/oxc_linter/src/rules/react/forbid_dom_props.rs:23` — Prop \
- [ ] `crates/oxc_linter/src/rules/react/jsx_key.rs:31` — r#"Missing "key" prop for element in array."#
- [ ] `crates/oxc_linter/src/rules/react/jsx_max_depth.rs:21` — format!(
        "JSX nesting depth of {depth} exceeds the configured maximum of {max}"
- [ ] `crates/oxc_linter/src/rules/react/jsx_no_comment_textnodes.rs:13` — Comments inside children section of tag should be placed inside braces
- [ ] `crates/oxc_linter/src/rules/react/jsx_no_undef.rs:16` — '{ident_name}' is not defined.
- [ ] `crates/oxc_linter/src/rules/react/jsx_no_useless_fragment.rs:24` — Fragments should contain more than one child.
- [ ] `crates/oxc_linter/src/rules/react/jsx_no_useless_fragment.rs:28` — Passing a fragment to a HTML element is useless.
- [ ] `crates/oxc_linter/src/rules/react/jsx_pascal_case.rs:26` — message
- [ ] `crates/oxc_linter/src/rules/react/jsx_props_no_spreading.rs:20` — Prop spreading is forbidden
- [ ] `crates/oxc_linter/src/rules/react/no_namespace.rs:16` — message
- [ ] `crates/oxc_linter/src/rules/react/no_redundant_should_component_update.rs:15` — format!(
        "{component_name} does not need `shouldComponentUpdate` when extending `React.PureComponent`."
- [ ] `crates/oxc_linter/src/rules/react/no_set_state.rs:14` — Do not use `setState`.
- [ ] `crates/oxc_linter/src/rules/react/no_unescaped_entities.rs:17` — `{unescaped}` can be escaped with {escaped}
- [ ] `crates/oxc_linter/src/rules/react/only_export_components.rs:16` — This rule can't verify that `export *` only exports components.
- [ ] `crates/oxc_linter/src/rules/react/only_export_components.rs:21` — Fast refresh only works when a file only exports components. Use a new file to share constants or functions between components.
- [ ] `crates/oxc_linter/src/rules/react/only_export_components.rs:26` — 
        "Fast refresh can't handle anonymous components. Add a name to your export.",
- [ ] `crates/oxc_linter/src/rules/react/only_export_components.rs:33` — Fast refresh only works when a file only exports components. Move your component(s) to a separate file.
- [ ] `crates/oxc_linter/src/rules/react/only_export_components.rs:38` — Fast refresh only works when a file has exports. Move your component(s) to a separate file.
- [ ] `crates/oxc_linter/src/rules/react/only_export_components.rs:43` — Fast refresh only works when a file only exports components. Move your React context(s) to a separate file.
- [ ] `crates/oxc_linter/src/rules/react/prefer_es6_class.rs:16` — Components should use `createReactClass` instead of an ES2015 class.
- [ ] `crates/oxc_linter/src/rules/react/prefer_es6_class.rs:21` — Components should use an ES2015 class instead of `createReactClass`.
- [ ] `crates/oxc_linter/src/rules/react/rules_of_hooks.rs:34` — format!(
            "React Hook {hook_name:?} is called in function {func_name:?} that is neither \
            a React function component nor a custom React Hook function. \
            React component names must start with an uppercase letter. \
            React Hook names must start with the word \"use\".",
- [ ] `crates/oxc_linter/src/rules/react/rules_of_hooks.rs:48` — format!(
            "React Hook {hook_name:?} is called conditionally. React Hooks must be \
            called in the exact same order in every component render."
- [ ] `crates/oxc_linter/src/rules/react/rules_of_hooks.rs:57` — format!(
            "React Hook {hook_name:?} may be executed more than once. Possibly \
            because it is called in a loop. React Hooks must be called in the \
            exact same order in every component render."
- [ ] `crates/oxc_linter/src/rules/react/rules_of_hooks.rs:67` — format!(
            "React Hook {hook_name:?} cannot be called at the top level. React Hooks \
            must be called in a React function component or a custom React \
            Hook function."
- [ ] `crates/oxc_linter/src/rules/react/rules_of_hooks.rs:77` — format!(
            "React Hook {func_name:?} cannot be called in an async function. "
- [ ] `crates/oxc_linter/src/rules/react/rules_of_hooks.rs:85` — format!(
            "React Hook {hook_name:?} cannot be called in a class component. React Hooks \
            must be called in a React function component or a custom React \
            Hook function."
- [ ] `crates/oxc_linter/src/rules/react/rules_of_hooks.rs:95` — format!(
            "React Hook {hook_name:?} cannot be called inside a callback. React Hooks \
            must be called in a React function component or a custom React \
            Hook function."
- [ ] `crates/oxc_linter/src/rules/react/state_in_constructor.rs:23` — message
- [ ] `crates/oxc_linter/src/rules/react/style_prop_object.rs:23` — `style` prop value must be an object.

</details>

### `typescript`

<details>

- [ ] `crates/oxc_linter/src/rules/typescript/adjacent_overload_signatures.rs:23` — All {fn_name:?} signatures should be adjacent.
- [ ] `crates/oxc_linter/src/rules/typescript/ban_ts_comment.rs:14` — format!(
        "Do not use @ts-{ts_comment_name} because it alters compilation errors."
- [ ] `crates/oxc_linter/src/rules/typescript/ban_ts_comment.rs:27` — format!(
        "Include a description after the @ts-{ts_comment_name} directive to explain why the @ts-{ts_comment_name} is necessary. The description must be {min_len} characters or longer."
- [ ] `crates/oxc_linter/src/rules/typescript/ban_ts_comment.rs:38` — format!(
        "The description for the @ts-{ts_comment_name} directive must match the {pattern} format."
- [ ] `crates/oxc_linter/src/rules/typescript/ban_types.rs:14` — format!(
        "Do not use {banned_type:?} as a type. Use \"{suggested_type}\" instead"
- [ ] `crates/oxc_linter/src/rules/typescript/ban_types.rs:33` — 'The `Object` type actually means \
- [ ] `crates/oxc_linter/src/rules/typescript/consistent_type_assertions.rs:22` — Use `<{cast}>` instead of `as {cast}`.
- [ ] `crates/oxc_linter/src/rules/typescript/consistent_type_assertions.rs:26` — Use `as {cast}` instead of `<{cast}>`.
- [ ] `crates/oxc_linter/src/rules/typescript/consistent_type_assertions.rs:30` — Do not use any type assertions.
- [ ] `crates/oxc_linter/src/rules/typescript/consistent_type_assertions.rs:34` — Always prefer `const x: T = { ... }`.
- [ ] `crates/oxc_linter/src/rules/typescript/consistent_type_assertions.rs:38` — Always prefer `const x: T[] = [ ... ]`.
- [ ] `crates/oxc_linter/src/rules/typescript/consistent_type_imports.rs:26` — `import()` type annotations are forbidden.
- [ ] `crates/oxc_linter/src/rules/typescript/consistent_type_imports.rs:30` — Use an `import` instead of an `import type`.
- [ ] `crates/oxc_linter/src/rules/typescript/consistent_type_imports.rs:33` — All imports in the declaration are only used as types. Use `import type`.
- [ ] `crates/oxc_linter/src/rules/typescript/consistent_type_imports.rs:38` — Imports {type_imports} are only used as type.
- [ ] `crates/oxc_linter/src/rules/typescript/explicit_module_boundary_types.rs:26` — Missing return type on function
- [ ] `crates/oxc_linter/src/rules/typescript/explicit_module_boundary_types.rs:30` — Missing argument type on function
- [ ] `crates/oxc_linter/src/rules/typescript/explicit_module_boundary_types.rs:34` — Argument is explicitly typed as `any`
- [ ] `crates/oxc_linter/src/rules/typescript/no_dynamic_delete.rs:42` — Do not delete dynamically computed property keys.
- [ ] `crates/oxc_linter/src/rules/typescript/no_empty_interface.rs:16` — an empty interface is equivalent to `{}`
- [ ] `crates/oxc_linter/src/rules/typescript/no_empty_interface.rs:20` — an interface declaring no members is equivalent to its supertype
- [ ] `crates/oxc_linter/src/rules/typescript/no_extra_non_null_assertion.rs:13` — extra non-null assertion
- [ ] `crates/oxc_linter/src/rules/typescript/no_wrapper_object_types.rs:13` — Do not use wrapper object types.
- [ ] `crates/oxc_linter/src/rules/typescript/prefer_enum_initializers.rs:14` — format!(
        "The value of the member {member_name:?} should be explicitly defined."
    
</details>

### `unicorn`

<details>

- [ ] `crates/oxc_linter/src/rules/unicorn/error_message.rs:12` — Pass a message to the {ctor_name:1} constructor.
- [ ] `crates/oxc_linter/src/rules/unicorn/error_message.rs:17` — Error message should not be an empty string.
- [ ] `crates/oxc_linter/src/rules/unicorn/error_message.rs:21` — Error message should be a string.
- [ ] `crates/oxc_linter/src/rules/unicorn/new_for_builtins.rs:10` — Use `new {fn_name}()` instead of `{fn_name}()`
- [ ] `crates/oxc_linter/src/rules/unicorn/new_for_builtins.rs:14` — Use `{fn_name}()` instead of `new {fn_name}()`
- [ ] `crates/oxc_linter/src/rules/unicorn/no_anonymous_default_export.rs:17` — This {kind} default export is missing a name
- [ ] `crates/oxc_linter/src/rules/unicorn/no_invalid_fetch_options.rs:21` — message
- [ ] `crates/oxc_linter/src/rules/unicorn/no_object_as_default_parameter.rs:12` — Do not use an object literal as default for parameter `{param}`.
- [ ] `crates/oxc_linter/src/rules/unicorn/no_object_as_default_parameter.rs:17` — Do not use an object literal as default
- [ ] `crates/oxc_linter/src/rules/unicorn/no_static_only_class.rs:9` — Use an object instead of a class with only static members.
- [ ] `crates/oxc_linter/src/rules/unicorn/no_typeof_undefined.rs:17` — Compare with `undefined` directly instead of using `typeof`.
- [ ] `crates/oxc_linter/src/rules/unicorn/no_unreadable_array_destructuring.rs:10` — Array destructuring may not contain consecutive ignored values.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_add_event_listener.rs:9` — Prefer `addEventListener()` over their `on`-function counterparts.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_array_some.rs:19` — Prefer `.some(…)` over `.find(…)` or `.findLast(…)`.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_array_some.rs:23` — Prefer `.some(…)` over non-zero length check from `.filter(…)`.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_array_some.rs:28` — Prefer `.some(…)` over `.findIndex(…)` or `.findLastIndex(…)`.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_blob_reading_methods.rs:13` — format!(
        "Prefer `Blob#{good_method}(
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_includes.rs:15` — 
        "Prefer `includes(
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_math_trunc.rs:10` — Prefer `Math.trunc()` over instead of `{bad_op} 0`.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_modern_dom_apis.rs:16` — Prefer using `{good_method}` over `{bad_method}`.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_modern_math_apis.rs:15` — Prefer `Math.abs(x)` over alternatives
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_modern_math_apis.rs:19` — Prefer `Math.hypot(…)` over alternatives
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_modern_math_apis.rs:23` — Prefer `Math.{good_method}(x)` over `{bad_method}`
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_native_coercion_functions.rs:13` — format!(
        "The function is equivalent to `{called_fn}`. Call `{called_fn}` directly."
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_native_coercion_functions.rs:20` — The arrow function in the callback of the array is equivalent to `Boolean`. Replace the callback with `Boolean`.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_prototype_methods.rs:15` — Prefer using `{obj_name}.prototype.{method_name}`.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_prototype_methods.rs:20` — Prefer using method from `{obj_name}.prototype`.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_response_static_json.rs:17` — Prefer using `Response.json(…)` over `JSON.stringify()`.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_string_starts_ends_with.rs:19` — Prefer String#startsWith over a regex with a caret.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_string_starts_ends_with.rs:23` — Prefer String#endsWith over a regex with a dollar sign.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_top_level_await.rs:12` — Prefer top-level await over using a promise chain.
- [ ] `crates/oxc_linter/src/rules/unicorn/prefer_top_level_await.rs:16` — Prefer top-level await over using an async IIFE.
- [ ] `crates/oxc_linter/src/rules/unicorn/relative_url_style.rs:20` — Remove the `./` prefix from the relative URL.
- [ ] `crates/oxc_linter/src/rules/unicorn/relative_url_style.rs:24` — Add a `./` prefix to the relative URL.

### `vue`

<details>

- [ ] `crates/oxc_linter/src/rules/vue/define_emits_declaration.rs:19` — Use type based declaration instead of runtime declaration
- [ ] `crates/oxc_linter/src/rules/vue/define_emits_declaration.rs:24` — Use runtime declaration instead of type based declaration
- [ ] `crates/oxc_linter/src/rules/vue/define_emits_declaration.rs:29` — 
        "Use new type literal declaration instead of the old call signature declaration",
- [ ] `crates/oxc_linter/src/rules/vue/define_props_declaration.rs:16` — Use runtime declaration instead of type-based declaration
- [ ] `crates/oxc_linter/src/rules/vue/define_props_declaration.rs:21` — Use type-based declaration instead of runtime declaration
- [ ] `crates/oxc_linter/src/rules/vue/define_props_destructuring.rs:16` — Prefer destructuring from `defineProps` directly.
- [ ] `crates/oxc_linter/src/rules/vue/define_props_destructuring.rs:20` — Avoid destructuring from `defineProps`.
- [ ] `crates/oxc_linter/src/rules/vue/define_props_destructuring.rs:24` — Avoid using `withDefaults` with destructuring.
- [ ] `crates/oxc_linter/src/rules/vue/no_export_in_script_setup.rs:12` — <script setup>` cannot contain ES module exports.
- [ ] `crates/oxc_linter/src/rules/vue/require_default_export.rs:13` — Missing default export.
- [ ] `crates/oxc_linter/src/rules/vue/require_default_export.rs:17` — Component must be the default export.

</details>

**Comment by anthonyamaro15** [comment]
Hi! Is this still open for contributions? I see a couple of PRs linked already, but it looks like there are plenty of diagnostics left to cover. If so, I'd like to start with some of the ESLint group, happy to pick up a handful and open a PR.

**Comment by camchenry** [comment]
> Hi! Is this still open for contributions? I see a couple of PRs linked already, but it looks like there are plenty of diagnostics left to cover. If so, I'd like to start with some of the ESLint group, happy to pick up a handful and open a PR.

Yes almost all of these are still available. I recommend splitting it up into small PRs with just a few (or even only one) rule updated, so it's more likely to get merged. Thanks for your interest!

**Comment by anthonyamaro15** [comment]
> > Hi! Is this still open for contributions? I see a couple of PRs linked already, but it looks like there are plenty of diagnostics left to cover. If so, I'd like to start with some of the ESLint group, happy to pick up a handful and open a PR.
> 
> Yes almost all of these are still available. I recommend splitting it up into small PRs with just a few (or even only one) rule updated, so it's more likely to get merged. Thanks for your interest!

Thank you!

**Comment by anthonyamaro15** [comment]
@camchenry, I went ahead and opened two PR's for this. For the second PR, I bundled  changes since the changes were all pretty much the same (just adding a
  `.with_help()` to each) But let me know if you'd rather I split them up!

**Comment by dosmond** [comment]
@camchenry I believe #19827 and #19826 are the last of the missing typescript diagnostics

**Comment by camc314** [comment]
Thank you to everyone that has been working on this.

Let me also add, if a rule has a fixer, the help message is usually automatically inserted by some internal logic - this help message shows exactly what should be changed to fix the issue so it's usually unnecessary to explicitly add a help message.

It's also worth being very considerate about the context in which the help message is being displayed in. For example. if the help message is pretty similar to the diagnostic message, it might be unnecessary to have the help message - in this case it might make sense to add additional context, explaining the issue in the note. With AI agents doing more and more of our work, it's important that we don't waste the tokens that are given to us!

**Comment by camchenry** [comment]
@camc314 I think we've addressed a lot of the easy help messages here. I haven't done a re-audit, but it seems like lots of progress has been made. What do you think about adding some sort of hint that diagnostics explicitly don't hand a help message added? Something like: `OxcDiagnostics::warn("...").with_label(...).no_help_message_needed()` where it only has an effect when debug assertions are enabled (should be compiled out) and just throws an error if there was a help message added to the diagnostic already.

**Comment by camchenry** [comment]
I think this is done for now. We made a lot of good progress, but the contributions coming in now are less helpful, which I think indicates we've done a lot of the easy wins here.
